Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,405 @@
+    }
+
+    /// - see: `EventLoop.scheduleTask(deadline:_:)`
+    @discardableResult
+    public func scheduleTask<T>(deadline: NIODeadline, _ task: @escaping () throws -> T) -> Scheduled<T> {
+        let promise: EventLoopPromise<T> = self.makePromise()
+        let taskID = self.scheduledTaskCounter.add(1)
+
+        let scheduled = Scheduled(promise: promise, cancellationTask: {
+            if self.inEventLoop {
+                self.removeTask(taskID: taskID)
+            } else {
+                self.queue.async {
+                    self.removeTask(taskID: taskID)
+                }
+            }
+        })
+
+        if self.inEventLoop {
+            self.insertTask(taskID: taskID, deadline: deadline, promise: promise, task: task)
+        } else {
+            self.queue.async {
+                self.insertTask(taskID: taskID, deadline: deadline, promise: promise, task: task)
+            }
+        }
+        return scheduled
+    }
+
+    /// - see: `EventLoop.scheduleTask(in:_:)`
+    @discardableResult
+    public func scheduleTask<T>(in: TimeAmount, _ task: @escaping () throws -> T) -> Scheduled<T> {
+        return self.scheduleTask(deadline: self.now + `in`, task)
+    }
+
+    /// On an `EmbeddedEventLoop`, `execute` will simply use `scheduleTask` with a deadline of _now_. This means that
+    /// `task` will be run the next time you call `EmbeddedEventLoop.run`.
+    public func execute(_ task: @escaping () -> Void) {
+        self.scheduleTask(deadline: self.now, task)
+    }
+
+    /// Run all tasks that have previously been submitted to this `EmbeddedEventLoop`, either by calling `execute` or
+    /// events that have been enqueued using `scheduleTask`/`scheduleRepeatedTask`/`scheduleRepeatedAsyncTask` and whose
+    /// deadlines have expired.
+    ///
+    /// - seealso: `EmbeddedEventLoop.advanceTime`.
+    public func run() async {
+        // Execute all tasks that are currently enqueued to be executed *now*.
+        await self.advanceTime(to: self.now)
+    }
+
+    /// Runs the event loop and moves "time" forward by the given amount, running any scheduled
+    /// tasks that need to be run.
+    public func advanceTime(by increment: TimeAmount) async {
+        await self.advanceTime(to: self.now + increment)
+    }
+
+    /// Unwrap a future result from this event loop.
+    ///
+    /// This replaces `EventLoopFuture.get()` for use with `AsyncEmbeddedEventLoop`. This is necessary because attaching
+    /// a callback to an `EventLoopFuture` (which is what `EventLoopFuture.get` does) requires scheduling a work item onto
+    /// the event loop, and running that work item requires spinning the loop. This is a non-trivial dance to get right due
+    /// to timing issues, so this function provides a helper to ensure that this will work.
+    public func awaitFuture<ResultType: Sendable>(_ future: EventLoopFuture<ResultType>, timeout: TimeAmount) async throws -> ResultType {
```

Comment:
We had a naming bikeshed here:

@dnadoba raised that `awaitFuture` will require us to write:
```swift
let foo = try await el.awaitFuture(future)
```
which has `await` twice and likely `future` twice, which is not nice to read.

---

@fabianfett proposed: `runUntil`, which IMO explicitly states the side-effects this method has.

@Lukasa didn't like `runUntil` and proposed `unwrap`.

Reply:
I definitely don't like most of these names.

Reply:
Reading through these suggestions I like `advanceTime(until:)` the most. IMO it is important that we convey that the loop is running when using this method.

Reply:
One issue with this name is that this method does not advance time. It just runs the loop repeatedly. We wouldn't know what time to advance the loop to.

Reply:
Since we should use this in favour of `EventLoopFuture.get()` what about just calling it `get`?

Reply:
Maybe the solution is to just compose `run` and `get` together like this:
```swift
try await eventLoop.run { 
    let foo = try await eventLoop.get()
    // do something with foo
}
// or 
let foo = try await eventLoop.run { try await eventLoop.get() }
// do something with foo
```
we could also call it `runUntil`. `runWhile` or something similar but the basic idea is that user needs to compose them and there is no special function that does both at the same time.
WDYT?

Reply:
The downside here is that I think this is less obvious than a single function. It relies on the user understanding _why_ a regular `wait` will hang, and being able to observe that this behaviour works. I think the `func run(_:) async` that this implies is also a very weird function, as it's essentially a "run until the other task completes". I can't think of another situation that requires this kind of pattern, so I'm inclined to lean away from it.